rpcs3/Emu/Cell/lv2/sys_fs.cpp

```diff
@@ -1649,15 +1662,79 @@ error_code sys_fs_fcntl(ppu_thread& ppu, u32 fd, u32 op, vm::ptr<void> _arg, u32
 
 	case 0xc0000007: // cellFsArcadeHddSerialNumber
 	{
-		const auto arg = vm::static_ptr_cast<lv2_file_c000007>(_arg);
+		const auto arg = vm::static_ptr_cast<lv2_file_c0000007>(_arg);
 		// TODO populate arg-> unk1+2
 		arg->out_code = CELL_OK;
 		return CELL_OK;
 	}
 
 	case 0xc0000008: // cellFsSetDefaultContainer, cellFsSetIoBuffer, cellFsSetIoBufferFromDefaultContainer
 	{
-		break;
+		// Allocates memory from a container/default container to a specific fd or default IO processing
+		const auto arg          = vm::static_ptr_cast<lv2_file_c0000008>(_arg);
+		auto& default_container = g_fxo->get<default_sys_fs_container>();
+
+		if (fd == 0xFFFFFFFF)
+		{
+			// No check on container is done when setting default container
+			default_container.id   = arg->size ? ::narrow<u32>(arg->container_id) : 0u;
+			default_container.cap  = arg->size;
+			default_container.used = 0;
+
+			arg->out_code = CELL_OK;
+			return CELL_OK;
+		}
+
+		auto file = idm::get<lv2_fs_object, lv2_file>(fd);
+		if (!file)
+		{
+			return CELL_EBADF;
+		}
+
+		auto free_current_container = [&]
+		{
+			if (auto container = idm::get<lv2_memory_container>(file->container_id))
+			{
+				container->used -= file->container_used;
+			}
+		};
+
+		if (!arg->size)
+		{
+			free_current_container();
+			file->container_id   = 0;
+			file->container_used = 0;
+			arg->out_code        = CELL_OK;
+			return CELL_OK;
+		}
+
+		// aligns on lower bound
+		u32 actual_size = arg->size - (arg->size % ((arg->page_type & CELL_FS_IO_BUFFER_PAGE_SIZE_64KB) ? 0x10000 : 0x100000));
+
+		if (arg->container_id == 0xFFFFFFFF && (default_container.used + actual_size) > default_container.cap)
+		{
+			return CELL_ENOMEM;
+		}
+
+		u32 new_container_id = arg->container_id == 0xFFFFFFFF ? default_container.id : ::narrow<u32>(arg->container_id);
+		auto new_container   = idm::get<lv2_memory_container>(new_container_id);
+
+		if (!new_container || (actual_size && !new_container->take(actual_size)))
```

Comment:
You need to take and free memory atomically, also if you take before freeing it may result in a false memory wrror.

Reply:
Not really resolved, I think it needs to operate under IDM mutex reader ownership such as by lambda and s_memstats_mtx writer ownership at sys_memory.cpp.

Reply:
Memory is freed before allocation is done.
Both take & free are atomic...I'm not sure why you want an extra mutex here?

Reply:
take is not supposed to operate outside of IDM mutex, and free may cause the container to be destructable and be destroyed by other thread by the time take takes place.
The second situation means that updating settings for the same container may fail because it has been destroyed by other thread inbetween.

Reply:
~~Complex threading headache, just do it :)~~